FastAPI: Histories API migration

davelopez opened the pull request:
This is a follow up on https://github.com/galaxyproject/galaxy/pull/12195 and part of #10889

This PR includes the new `FastAPIHistories` controller that replaces the old legacy routes for [galaxy/api/histories.py](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/api/histories.py) and some minor additional refactorings and types annotations,

This is still a WIP as there are some fundamental issues that need to be addressed:
- [x] The `create` action needs to be adapted to accept both, a file and body in the request. Looks like the way to go is by adding a new dependency as per the [FastAPI docs](https://fastapi.tiangolo.com/tutorial/request-forms-and-files/) <del>but still need to figure out how to do it</del>. Done in https://github.com/galaxyproject/galaxy/pull/12386/commits/dbaa4d8b58f59f57bd9c790e9d8dcf93981513f1.
- [x] The handling of the file response in `archive_download` needs to be implemented in a different way in FastAPI. Done in https://github.com/galaxyproject/galaxy/pull/12386/commits/2e98dffd6fac3189ee1ba6e20f15f1b0d594ddd2.
- [x] When exporting a history, we need to generate proper links to download it<del>, but [we can't use `url_for` for that](https://github.com/galaxyproject/galaxy/pull/12370). I tried to play with the solution in https://github.com/galaxyproject/galaxy/pull/11904 but couldn't make it work, there is probably something wrong is how the FastAPI routers are set up or the way the function should be called but I always get a `NoMatchFound`. I need to investigate further on this one.</del> Should be possible with https://github.com/galaxyproject/galaxy/pull/12391. See https://github.com/galaxyproject/galaxy/pull/12386#issuecomment-910301555 for details.
- [x] The `run_as` feature is not working with FastAPI but fortunately the existing tests on `lib/galaxy_test/api/test_framework.py` caught it. Fixed, more details in https://github.com/galaxyproject/galaxy/pull/12386#issuecomment-910392728.
- [x] There is something wrong with access to histories as an anonymous user, investigating... **Update**: apparently, the `current_history` was never set in the `SessionRequestContext` so anonymous users couldn't access the default history. For more details see https://github.com/galaxyproject/galaxy/pull/12386#discussion_r701178647.
- [x] There is still an issue with the `create` action. For more details see https://github.com/galaxyproject/galaxy/pull/12386#issuecomment-911922465. Fixed in https://github.com/galaxyproject/galaxy/pull/12386/commits/657bb8bc1f2edf46c23aab6b3171a37ed2df6dcb.
- [x] One last legitimate [integration test is failing](https://github.com/galaxyproject/galaxy/pull/12386/checks?check_run_id=3553852292), investigating... **Update**: I unintentionally broke the condition for returning 204 status code in `archive_export`, that was a bit hard to notice but should be fixed by https://github.com/galaxyproject/galaxy/pull/12386/commits/b86990fe3a5788056a24b0488e7173698642af08.

## How to test the changes?
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Go to the interactive API documentation at `http://127.0.0.1:8080/api/docs` and play with it :)
    ![Screenshot from 2021-08-27 14-45-57](https://user-images.githubusercontent.com/46503462/131129376-73a95e19-527c-4cb9-938c-aa78c4b4b8b8.png)


## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Refactor SerializationParams dict into a model

davelopez committed:
Add explicit exception when trying to create a history as anonymous

This way it will also work in FastAPI.
Also fixed a couple of type hints.

davelopez committed:
Adapt test to use JSON headers

Required by FastAPI, see comment https://github.com/galaxyproject/galaxy/pull/12152#issuecomment-864134487

davelopez committed:
Add FastAPI routes for histories API

davelopez committed:
Move response status to API level

The 204 status in `archive_export` should be changed at the API level and not the service level, especially while both, legacy and FastAPI controllers coexist as the response is handled differently.

davelopez committed:
Handle `archive_download` action

The implementation for the FastAPI version of the endpoint differs from the legacy one, so it cannot be totally reused like in the rest of the routes.

The response is now treated as a Streaming response and the metadata is handled at the API level instead of the manager.

davelopez committed:
Handle `create` action using form data

The create action involves file uploading for history import, so we have to use Form data instead of JSON. See docs https://fastapi.tiangolo.com/tutorial/request-forms-and-files/
- Add required `python-multipart` dependency
- Add decorator `as_form` that allows to use a pydantic model as form-data
- Handle how FastAPI deals with uploaded files using SpooledTemporaryFile

davelopez committed:
Adapt test_parse_serialization_params to expect a model

davelopez committed:
Fix mypy

davelopez committed:
Fix double dict conversion in configuration API

This was unintentionally introduced in 15e69de380a01ed8bef6b1c9d12dbbf6b042d5ae

davelopez committed:
Generate API URLs for downloading history exports

This may not be the best way to generate the URLs since this is forcing all to be API URLs but I couldn't think of a better solution to make it work for both legacy and FastAPI frameworks.

davelopez committed:
Fix `run_as` to be passed in the headers

The legacy API framework allowed to pass the `run_as` user in the payload but, moving forward to the new framework, looks like the headers is a better place for it (see https://github.com/galaxyproject/galaxy/pull/11342/commits/91e7d5191d70f3b286fdf86b39af657bede35f09).

- Adapt existing framework tests to pass in the `run_as` user in the headers instead of the payload.
- Fix the legacy framework not retrieving the `run_as` from the headers when the payload was not in JSON (i.e. Form data)
- Add a bit more of documentation for the OpenAPI schema.

davelopez committed:
Fix `current_history` not set in `SessionRequestContext`

davelopez committed:
Add `Any` as a valid model for History view

This is the only way I can think of to support arbitrary serialization `keys` being requested. Otherwise, the validation will fail because the required fields in the annotated models may not be present in the resulting `custom` serialized history.

davelopez commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -150,11 +150,22 @@ def __call__(self, name: str, **path_params):
+        app=app, user=user,
+        galaxy_session=galaxy_session,
+        url_builder=url_builder, host=request.client.host,
+        history=get_current_history_from_session(galaxy_session),
```
The `current_history` seemed to be always None when using Galaxy as an anonymous user in FastAPI mode.

Hopefully getting it from the session is the correct way to set it, right?

mvdbeek replied:
Yes, I think that looks right

davelopez committed:
Add support for JSON payload in `create` operation + test

davelopez committed:
Fix 204 response status condition on `archive_export`

mvdbeek commented on lib/galaxy/webapps/galaxy/api/histories.py:
```diff
@@ -28,16 +63,338 @@
+        # We will assume x-www-form-urlencoded by default to deal with possible file uploads.
+        # See https://github.com/tiangolo/fastapi/issues/990#issuecomment-639615888
+        if "application/json" in request.headers.get("content-type", ""):
+            body = await request.json()
```
Is it mandatory to use await here ? According to the docs we shouldn't create async routes as long as we do synchronous IO (which we do by creating the history in the database) ... not quite sure what the implications are since the sync code runs in a threadpool, but if it is possible to make this a sync method I think we should do that.

mvdbeek replied:
And if we need to use `await`, could we make this an async path dependency ?

davelopez replied:
Good point! It is mandatory to use await as the `request.json()` function returns a coroutine, but I didn't think about using an async path dependency, so thank you very much for the hint!

davelopez committed:
Make create path operation sync

We need to use `await request.json()` because that function returns a coroutine, but we can move this into an async path dependency as Marius suggested so the path operation remains sync.

mvdbeek commented on lib/galaxy/managers/histories.py:
```diff
@@ -221,6 +232,16 @@ def serve_ready_history_export(self, trans, jeha):
 
+    def serve_ready_history_export(self, trans, jeha):
+        """
+        Serves the history export archive for use as a streaming response so the file
```
The file is not loaded into memory with the legacy method either, and the file handle is passed on to nginx, completely freeing the handler. I think that is preferable over a streaming response, even if it is async. Is there a downside I am not thinking of with what is now the `legacy_serve_ready_history_export` method that would prevent using it?

davelopez replied:
That's a good question :)
I just read the documentation on how to deal with files in FastAPI and thought that [StreamingResponse](https://fastapi.tiangolo.com/advanced/custom-response/#streamingresponse) was a better fit than [FileResponse](https://fastapi.tiangolo.com/advanced/custom-response/#fileresponse) because we already had a file-like handle in `serve_ready_history_export`, but honestly, I have no idea which option is best as both seems to be streaming the file asynchronously in the end.

This has just reminded me of https://github.com/galaxyproject/galaxy/pull/11167#discussion_r559535763 which I promised to do :)

Do you think the way to go here is using a FileResponse but introducing the "send file middleware" to set the headers? Or what do you think should be the best approach?

mvdbeek replied:
> Do you think the way to go here is using a FileResponse but introducing the "send file middleware" to set the headers?

Yeah, I think that sounds good to me. If you set the middleware here you also have access to the config: https://github.com/mvdbeek/galaxy/blob/611fb8fe8b3a1994e97d7399a6aa8202bffe7a77/lib/galaxy/webapps/galaxy/fast_app.py#L61

davelopez replied:
Cool! Thanks! I will try!

davelopez replied:
Done!, I hope this will work https://github.com/galaxyproject/galaxy/pull/12441, don't really know how to properly test it though.

mvdbeek replied:
Alright, then I think what's left to do is to use the FileResponse ?

davelopez replied:
Done in https://github.com/galaxyproject/galaxy/pull/12386/commits/29db23928a5e3937575c92e24922fc91dd930186 :)

jmchilton commented on lib/galaxy/managers/histories.py:
```diff
@@ -955,28 +1028,33 @@ def get_custom_builds_metadata(self, trans, id: EncodedDatabaseIdField) -> Custo
             serialization_params: SerializationParams,
             default_view: str = "detailed",
-    ) -> Union[HistoryBeta, HistoryDetailed, HistorySummary]:
+    ) -> Union[AnyHistoryView]:
```
This is already a union right? Should this just be ``AnyHistoryView``?

davelopez replied:
Good catch! that's just a bad replace operation :sweat_smile:

davelopez committed:
Remove redundant Union

davelopez committed:
Use FileResponse instead of StreamingResponse
